### PR TITLE
Use `ExternalAddr` attribute for inter-node communication

### PR DESCRIPTION
**Pull request by fyrchik**

### CODE REVIEW

**Review by acid-ant — APPROVED** [review]

File: pkg/core/netmap/nodes.go [hunk 1]
```diff
@@ -17,12 +17,15 @@ func (x Node) PublicKey() []byte {
 // IterateAddresses iterates over all announced network addresses
 // and passes them into f. Handler MUST NOT be nil.
 func (x Node) IterateAddresses(f func(string) bool) {
+	for _, addr := range (netmap.NodeInfo)(x).ExternalAddresses() {
```
Comment (cthulhu-rider) [inline pkg/core/netmap/nodes.go:20]: So firstly we are going to process external addresses everywhere, is it expected for intra-container communication? I thought "old" addresses are preferred for this purpose.

**Comment by cthulhu-rider** [comment]
@fyrchik are you sure you've changed all the needed places? I see a few more places of `NumberOfNetworkEndpoints` and `IterateNetworkEndpoints`.

`pkg/core/netmap.Node` type was borned to become a complete type mediator for SDK `netmap.NodeInfo`, but we haven't finished the job yet. Lets do it within a separate issue?

**Comment by realloc** [comment]
IIRC, the ExternalAddr attribute should be used for client (sdk?) connections to nodes. For inter-node communication we have to use addressed from netmap and use ExternalAddr as a fallback only, if there is an option allowing such behaviour.

**Comment by fyrchik** [comment]
There are 3 different approaches:
1. Don't touch the address structure, and decide if we want to use external addressses on the callsite (client cache).
2. Modify network.Address to have `external` field and take it into account during sorting.
3. Filter external addresses during `placement.Node` construction.

I decided to go with the first, because this way there are fewer places with filtering logic, so less bugs. Also, it will be easier to change `allow_external` setting on SIGHUP.

File: CHANGELOG.md [hunk 2]
```diff
@@ -6,7 +6,7 @@ Changelog for NeoFS Node
 ### Added
 - Serving `NetmapService.NetmapSnapshot` RPC (#1793)
 - `netmap snapshot` command of NeoFS CLI (#1793)
-
+- `apiclient.allow_external` config flag to fallback to node external addresses (#1817)
```
Comment (carpawell) [inline CHANGELOG.md:9]: do we need to add that to the `Update from...` section too?

File: pkg/core/client/client.go [hunk 3]
```diff
@@ -52,6 +54,16 @@ func (x NodeInfo) AddressGroup() network.AddressGroup {
 	return x.addrGroup
 }
 
+// SetExternalAddressGroup sets a group of network addresses.
```
Comment (carpawell) [inline pkg/core/client/client.go:57]: should it have _the same_ comment as `SetAddressGroup`?

File: pkg/core/netmap/nodes.go [hunk 4]
```diff
@@ -17,6 +17,9 @@ func (x Node) PublicKey() []byte {
...
+		f(addr)
```
Comment (carpawell) [inline pkg/core/netmap/nodes.go:21]: no break if `f(addr) == true`?

File: CHANGELOG.md (hunk 2 above)
Comment (fyrchik) [reply CHANGELOG.md:9]: I think no, it is a completely new feature, old nodes are expected to continue working.

File: pkg/core/client/client.go (hunk 3 above)
Comment (fyrchik) [reply pkg/core/client/client.go]: Fixed

File: pkg/core/netmap/nodes.go (hunk 4 above)
Comment (fyrchik) [reply pkg/core/netmap/nodes.go]: Fixed

File: CHANGELOG.md (hunk 2 above)
Comment (carpawell) [reply CHANGELOG.md:9]: yes, but how will an admin know about that? i try to write any config changes to that section. but ok

**Review by carpawell — DISMISSED** [review]

File: CHANGELOG.md (hunk 2 above)
Comment (realloc) [reply CHANGELOG.md:9]: I suggest adding this to Update section. With new behaviour node admins can explicitly declare external addresses for their node, not only listen corresponding addresses.

Comment (fyrchik) [reply CHANGELOG.md:9]: Yes, but do we need `Added` section in this case? Every added feature can be used in some way, I've though `Updating from` was for things requiring special care.
Fixed anyway.

Comment (realloc) [reply CHANGELOG.md:9]: If we need Ops attention, we add/duplicate something in `Updating from`. For this particular case we need to highlight it.

**Review by carpawell — APPROVED** [review]

**Pull request merged**